### PR TITLE
feat: App grouping build and deploy page

### PR DESCRIPTION
**Pull request by vivek-devtron**
# Description

This PR consist the changes of bulk CI and CD trigger

Fixes #[AB#1950](https://dev.azure.com/DevtronLabs/Devtron/_workitems/edit/1950)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?
Tested by creating multiple type of pipelines
- triggered bulk CI
- triggered bulk CD
- triggered single app flow

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/app/details/triggerView/ciMaterial.tsx [hunk 1]
```diff
@@ -158,6 +158,10 @@ export class CIMaterial extends Component<CIMaterialProps, CIMaterialState> {
                         isWebhookPayloadLoading={this.props.isWebhookPayloadLoading}
                         workflowId={this.props.workflowId}
                         onClickShowBranchRegexModal={this.props.onClickShowBranchRegexModal}
+                        isFromEnv={this.props.isFromEnv}
+                        appId={this.props.appId}
+                        isFromBulkCI={false}
+                        isHideSearchHeader={false}
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/ciMaterial.tsx:164]: Can we rename it to `hideSearchHeader`?

File: src/css/triggerView.css [hunk 2]
```diff
@@ -6,15 +6,9 @@
...
-    font-style: normal;
-    font-stretch: normal;
-    line-height: 1.4;
-    letter-spacing: normal;
-    color: var(--N700);
-    margin-top: 12px;
     display: flex;
-    align-items: flex-end;
+    justify-content: center;
+    align-items: center;
```
Comment (sohel-rp) [inline src/css/triggerView.css:11]: can we use predefined classes instead?

File: src/components/deploymentGroups/BulkActionDetails.tsx [hunk 3]
```diff
@@ -216,12 +217,12 @@ export default function BulkActionDetails() {
...
             isLoading={isLoading}
             envName={""}
-            stageType="CD"
+            stageType={DeploymentNodeType.CD}
             materialType="none"
             triggerDeploy={triggerDeploy}
             toggleSourceInfo={toggleSourceInfo}
             selectImage={selectImage}
-            closeCDModal={() => { toggleShowCDModal(false) }}
+            closeCDModal={(e) => { toggleShowCDModal(false) }}
```
Comment (sohel-rp) [inline src/components/deploymentGroups/BulkActionDetails.tsx:225]: Extract this

File: src/components/app/details/triggerView/workflow/nodes/triggerPrePostCDNode.tsx [hunk 4]
```diff
@@ -18,18 +18,29 @@ export class TriggerPrePostCDNode extends Component<TriggerPrePostCDNodeProps>{
     }
 
     redirectToCDDetails(e) {
-        const LINK = this.getCDDetailsURL();
-        this.props.history.push(LINK);
+      if (this.props.isFromEnv) {
+          return
+      }
+      const LINK = this.getCDDetailsURL()
+      this.props.history.push(LINK)
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/workflow/nodes/triggerPrePostCDNode.tsx:25]: Use `getCDDetailsURL` value directly

File: src/components/app/details/triggerView/workflow/nodes/triggerPrePostCDNode.tsx [hunk 5]
```diff
@@ -18,18 +18,29 @@ export class TriggerPrePostCDNode extends Component<TriggerPrePostCDNodeProps>{
...
-                <span>{this.props.status}</span>
-                <span className="mr-5 ml-5">/</span>
-                <Link to={url} className="workflow-node__details-link">Details</Link>
-            </div>
+            return (
+                <div className="dc__cd-trigger-status" style={{ color: TriggerStatus[status] }}>
+                    <span>{this.props.status}</span>
+                    {!this.props.isFromEnv && (
+                        <>
+                            <span className="mr-5 ml-5">/</span>
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/workflow/nodes/triggerPrePostCDNode.tsx:36]: Do we need this? If yes then change it to self closing tag.

File: src/components/app/details/triggerView/workflow/nodes/triggerCINode.tsx [hunk 6]
```diff
@@ -40,8 +41,11 @@ export class TriggerCINode extends Component<TriggerCINodeProps> {
     }
 
     redirectToCIDetails() {
-        const LINK = this.getCIDetailsURL()
-        this.props.history.push(LINK)
+      if (this.props.isFromEnv) {
+          return
+      }
+      const LINK = this.getCIDetailsURL()
+      this.props.history.push(LINK)
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/workflow/nodes/triggerCINode.tsx:48]: same here

File: src/components/app/details/triggerView/workflow/nodes/triggerCDNode.tsx [hunk 7]
```diff
@@ -91,7 +98,7 @@ export class TriggerCDNode extends Component<TriggerCDNodeProps> {
                                     className="workflow-node__deploy-btn"
                                     onClick={(event) => {
                                         event.stopPropagation()
-                                        context.onClickCDMaterial(this.props.id, this.props.type)
+                                        context.onClickCDMaterial(this.props.id, DeploymentNodeType[this.props.type])
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/workflow/nodes/triggerCDNode.tsx:101]: Extract onClick function as props values are used directly

File: src/components/app/details/triggerView/workflow/nodes/TriggerLinkedCINode.tsx [hunk 8]
```diff
@@ -33,8 +34,11 @@ export class TriggerLinkedCINode extends Component<CINodeProps> {
     }
 
     redirectToCIDetails() {
-        const LINK = this.getCIDetailsURL();
-        this.props.history.push(LINK);
+      if (this.props.isFromEnv) {
+          return
+      }
+      const LINK = this.getCIDetailsURL()
+      this.props.history.push(LINK)
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/workflow/nodes/TriggerLinkedCINode.tsx:41]: same here

File: src/components/Environments/Environment.service.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+export const getWorkflows = (envID): Promise<{ workflows: WorkflowType[]; filteredCIPipelines }> => {
+    const _workflows: WorkflowType[] = []
+    const _filteredCIPipelines = new Map()
+    return Promise.all([
+        getEnvWorkflowList(envID),
+        getCIConfig(envID),
+        getCDConfig(envID),
+        getExternalCIList(envID),
+    ]).then(([workflow, ciConfig, cdConfig, externalCIConfig]) => {
+        let _ciConfigMap = new Map<number, CiPipelineResult>()
```
Comment (sohel-rp) [inline src/components/Environments/Environment.service.ts:49]: Make it const

File: src/components/Environments/Environment.service.ts [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+export function getCIConfig(envID: number): Promise<ResponseType> {
+    const URL = `${Routes.ENV_WORKFLOW}/${envID}/${URLS.APP_CI_CONFIG}`
+    return get(URL)
+}
+
+export function getCDConfig(envID: number | string): Promise<ResponseType> {
+    const URL = `${Routes.ENV_WORKFLOW}/${envID}/${URLS.APP_CD_CONFIG}`
+    return get(URL)
```
Comment (sohel-rp) [inline src/components/Environments/Environment.service.ts:29]: Use URL directly in all above functions

File: src/components/Environments/Environment.service.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+export function getExternalCIList(envID: number | string): Promise<WebhookListResponse> {
+    return get(`${Routes.ENV_WORKFLOW}/${envID}/${URLS.APP_EXTERNAL_CI_CONFIG}`)
+}
+
+export const getWorkflowStatus = (envID: string) => {
+    return get(`${Routes.ENV_WORKFLOW}/${envID}/${Routes.WORKFLOW_STATUS}`)
+}
+
+export const getWorkflows = (envID): Promise<{ workflows: WorkflowType[]; filteredCIPipelines }> => {
```
Comment (sohel-rp) [inline src/components/Environments/Environment.service.ts:40]: move to types & provide arg a type

Comment (sohel-rp) [reply src/components/Environments/Environment.service.ts:40]: Also, can be changed to `envId`

File: src/components/Environments/Environment.service.ts [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+                filterChildAndSiblingCD(envID),
+            )
+            _workflows.push(...processWorkflowData.workflows)
+            _filteredCIPipelines.set(workflowResult.appId, processWorkflowData.filteredCIPipelines)
+        }
+        return { workflows: _workflows, filteredCIPipelines: _filteredCIPipelines }
+    })
+}
+
+const filterChildAndSiblingCD = function (envID: number): (workflows: WorkflowType[]) => WorkflowType[] {
```
Comment (sohel-rp) [inline src/components/Environments/Environment.service.ts:73]: same here

File: src/components/Environments/Environment.service.ts [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    return get(`${Routes.ENV_WORKFLOW}/${envID}/${URLS.APP_EXTERNAL_CI_CONFIG}`)
+}
+
+export const getWorkflowStatus = (envID: string) => {
+    return get(`${Routes.ENV_WORKFLOW}/${envID}/${Routes.WORKFLOW_STATUS}`)
+}
+
+export const getWorkflows = (envID): Promise<{ workflows: WorkflowType[]; filteredCIPipelines }> => {
+    const _workflows: WorkflowType[] = []
+    const _filteredCIPipelines = new Map()
```
Comment (sohel-rp) [inline src/components/Environments/Environment.service.ts:42]: same here

File: src/components/Environments/Environment.service.ts [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+            _filteredCIPipelines.set(workflowResult.appId, processWorkflowData.filteredCIPipelines)
+        }
+        return { workflows: _workflows, filteredCIPipelines: _filteredCIPipelines }
+    })
+}
+
+const filterChildAndSiblingCD = function (envID: number): (workflows: WorkflowType[]) => WorkflowType[] {
+    return (workflows: WorkflowType[]): WorkflowType[] => {
+        workflows.forEach((wf) => {
+            let nodes = new Map(wf.nodes.map((node) => [node.type + '-' + node.id, node] as [string, NodeAttr]))
```
Comment (sohel-rp) [inline src/components/Environments/Environment.service.ts:76]: make it const

Comment (sohel-rp) [reply src/components/Environments/Environment.service.ts:76]: Also remove commented

File: src/components/Environments/Environment.service.ts [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    } else if (node.parentPipelineType == PipelineType.WEBHOOK) {
+        parentType = WorkflowNodeType.WEBHOOK
+    }
+
+    let parentNode = nodes.get(parentType + '-' + node.parentPipelineId)
+
+    const type = node.preNode ? WorkflowNodeType.PRE_CD : node.type
+
+    if (!!parentNode) {
+        ;(parentNode.postNode ? parentNode.postNode : parentNode).downstreams = [type + '-' + node.id]
```
Comment (sohel-rp) [inline src/components/Environments/Environment.service.ts:116]: fix the formatting

File: src/components/Environments/Details/TriggerView/BulkCDTrigger.tsx [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+    const onClickStartDeploy = (): void => {
+      onClickTriggerBulkCD()
+    }
+
+    const renderFooterSection = (): JSX.Element => {
+        return (
+            <div
+                className="dc__border-top flex right bcn-0 pt-16 pr-20 pb-16 pl-20 dc__position-fixed dc__bottom-0"
+                style={{ width: '75%', minWidth: '1024px', maxWidth: '1200px' }}
```
Comment (sohel-rp) [inline src/components/Environments/Details/TriggerView/BulkCDTrigger.tsx:167]: class can be used

File: src/components/Environments/Details/TriggerView/TriggerResponseModal.tsx [hunk 17]
```diff
@@ -0,0 +1,92 @@
+import React from 'react'
+import { Progressing } from '../../../common'
+import { ReactComponent as RetryIcon } from '../../../../assets/icons/ic-arrow-clockwise.svg'
+import { ReactComponent as Success } from '../../../../assets/icons/appstatus/healthy.svg'
+import { ReactComponent as Error } from '../../../../assets/icons/ic-error-exclamation.svg'
+import { ReactComponent as UnAuthorized } from '../../../../assets/icons/ic-locked.svg'
+import { ResponseRowType } from '../../Environments.types'
+import { BulkResponseStatus } from '../../Constants'
+
+export default function TriggerResponseModal({ closePopup, responseList, isLoading, onClickRetryBuild }) {
```
Comment (sohel-rp) [inline src/components/Environments/Details/TriggerView/TriggerResponseModal.tsx:10]: same here

File: src/components/Environments/Details/TriggerView/TriggerResponseModal.tsx [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        onClickRetryBuild(appsToRetry)
+    }
+
+    const renderFooterSection = (): JSX.Element => {
+        return (
+            <div
+                className={`dc__border-top flex bcn-0 pt-16 pr-20 pb-16 pl-20 dc__position-fixed dc__bottom-0 ${
+                    isShowRetryButton ? 'dc__content-space' : 'right'
+                }`}
+                style={{ width: '75%', minWidth: '1024px', maxWidth: '1200px' }}
```
Comment (sohel-rp) [inline src/components/Environments/Details/TriggerView/TriggerResponseModal.tsx:65]: same here

File: src/components/Environments/Details/TriggerView/BulkCITrigger.tsx [hunk 19]
```diff
@@ -0,0 +1,535 @@
...
+                </>
+            )
+        } else if (selectedApp.isLinkedCI) {
+            return (
+                <EmptyView
+                    imgSrc={linkedCiImg}
+                    title={`${selectedApp.name} is using a linked build pipeline`}
+                    subTitle="You can trigger the parent build pipeline. Triggering the parent build pipeline will trigger all build pipelines linked to it."
+                    link={`${URLS.APP}/${selectedApp.parentAppId}/${URLS.APP_CI_DETAILS}/${selectedApp.parentCIPipelineId}`}
+                    linkText="View Source Pipeline"
```
Comment (sohel-rp) [inline src/components/Environments/Details/TriggerView/BulkCITrigger.tsx:278]: Move to const

File: src/components/Environments/Details/TriggerView/BulkCITrigger.tsx [hunk 20]
```diff
@@ -0,0 +1,535 @@
...
+                />
+            )
+        } else if (selectedApp.isWebhookCI) {
+            return (
+                <EmptyView
+                    imgSrc={externalCiImg}
+                    title={`${selectedApp.name} is using a external build pipeline`}
+                    subTitle="Images received from the external service will be available for deployment."
```
Comment (sohel-rp) [inline src/components/Environments/Details/TriggerView/BulkCITrigger.tsx:286]: same here

File: src/components/Environments/Details/TriggerView/BulkCITrigger.tsx [hunk 21]
```diff
@@ -0,0 +1,535 @@
...
+                    <InfoIcon className="icon-dim-20 info-icon-n6 mr-4" />
+                    <span className="fw-4 fs-13 cn-9">{infoText}</span>
+                </div>
+            </Tippy>
+        )
+    }
+
+    const renderCacheSection = (): JSX.Element | null => {
+        if (!selectedApp.isLinkedCI && !selectedApp.isWebhookCI && !showRegexModal) {
+            if (selectedApp.isFirstTrigger) {
```
Comment (sohel-rp) [inline src/components/Environments/Details/TriggerView/BulkCITrigger.tsx:347]: same here move messaging to const

File: src/components/Environments/Details/TriggerView/BulkCITrigger.tsx [hunk 22]
```diff
@@ -0,0 +1,535 @@
...
+
+    const renderFooterSection = (): JSX.Element => {
+        return (
+            <div
+                className={`dc__border-top flex right bcn-0 pt-16 pr-20 pb-16 pl-20 dc__position-fixed dc__bottom-0 ${
+                    !blobStorageConfigurationLoading && !blobStorageConfiguration?.result?.enabled
+                        ? 'dc__content-space'
+                        : ''
+                }`}
+                style={{ width: '75%', minWidth: '1024px', maxWidth: '1200px' }}
```
Comment (sohel-rp) [inline src/components/Environments/Details/TriggerView/BulkCITrigger.tsx:480]: same here

File: src/components/Environments/Environment.service.ts (hunk 14 above)
Comment (vivek-devtron) [reply src/components/Environments/Environment.service.ts:76]: It is re assigned below

Comment (sohel-rp) [reply src/components/Environments/Environment.service.ts:74]: Looks like `wf.nodes` is getting re-assigned. Check & update.

File: src/components/Environments/Details/TriggerView/EnvTriggerView.tsx [hunk 23]
```diff
@@ -0,0 +1,1500 @@
...
+                    }
+                    if (!element.commit) {
+                        historyItem.GitCommit['WebhookData'] = {
+                            id: element.webhookData.id,
+                        }
+                    }
+                    ciPipelineMaterials.push(historyItem)
+                })
+            }
+            // if (gitMaterials[dockerfileConfiguredGitMaterialId][1] === DEFAULT_GIT_BRANCH_VALUE) {
```
Comment (sohel-rp) [inline src/components/Environments/Details/TriggerView/EnvTriggerView.tsx:1068]: remove commented

**Pull request merged**